src/main/scala/org/broadinstitute/dsde/snoop/ws/PerRequest.scala

```diff
@@ -0,0 +1,79 @@
+package org.broadinstitute.dsde.snoop.ws
+
+import akka.actor._
+import akka.actor.SupervisorStrategy.Stop
+import org.broadinstitute.dsde.snoop.ws.PerRequest.{WithProps, WithActorRef}
+import spray.http.StatusCodes._
+import spray.routing.RequestContext
+import akka.actor.OneForOneStrategy
+import spray.httpx.Json4sSupport
+import scala.concurrent.duration._
+import org.json4s.DefaultFormats
+import spray.http.StatusCode
+
+/**
+ * This actor controls the lifecycle of a request. It is responsible for forwarding the initial message
+ * to a target handling actor. This actor waits for the target actor signal completion (via a message),
+ * timeout, or handle an exception. It is this actors responsibility to respond to the request and
+ * shutdown itself and child actors.
+ */
+trait PerRequest extends Actor with Json4sSupport {
+
+  import context._
+
+  val json4sFormats = DefaultFormats
+
+  def r: RequestContext
+  def target: ActorRef
+  def message: RestMessage
+  def timeout: Duration
+
+  setReceiveTimeout(timeout)
+  target ! message
+
+  def receive = {
+    case res: RestMessage => complete(OK, res)
+    case v: Validation    => complete(BadRequest, v)
+    case ReceiveTimeout   => complete(GatewayTimeout, "Request timeout")
+  }
+
+  def complete[T <: AnyRef](status: StatusCode, obj: T) = {
+    r.complete(status, obj)
+    stop(self)
+  }
+
+  override val supervisorStrategy =
+    OneForOneStrategy() {
+      case e => {
+        complete(InternalServerError, e.getMessage)
+        Stop
+      }
+    }
+}
+
+object PerRequest {
+  case class WithActorRef(r: RequestContext, target: ActorRef, message: RestMessage, timeout: Duration) extends PerRequest
+
+  case class WithProps(r: RequestContext, props: Props, message: RestMessage, timeout: Duration) extends PerRequest {
+    lazy val target = context.actorOf(props, "foo")
+  }
+}
+
+/**
+ * Provides factory methods for creating per request actors
+ */
+trait PerRequestCreator {
+  implicit def actorRefFactory: ActorRefFactory
+
+  def perRequest(r: RequestContext, target: ActorRef, message: RestMessage) =
```

Comment:
Cause I copied this code from NET-A-PORTER.

But this is supposed to be infrastructury so it seemed to make sense to keep these factory methods in case we want them later...